cloudinit/sources/DataSourceWSL.py

```diff
@@ -0,0 +1,234 @@
+    in some supported format, ordered by precedence.
+    """
+    distribution_id, _, release_codename = util.get_linux_distro()
+
+    return [
+        # WSL instance specific:
+        "%s.user-data" % instance_name,
+        # release codename specific
+        "%s-%s.user-data" % (distribution_id, release_codename),
+        # distribution specific (Alpine, Arch, Fedora, openSUSE, Ubuntu...)
+        "%s-all.user-data" % distribution_id,
+        # generic, valid for all WSL distros and instances.
+        "config.user-data",
+    ]
+
+
+class DataSourceWSL(sources.DataSource):
+    dsname = "WSL"
+
+    def __init__(self, sys_cfg, distro, paths):
+        sources.DataSource.__init__(self, sys_cfg, distro, paths)
+        self._network_config = sources.UNSET
+        self.dsmode = sources.DSMODE_LOCAL
+        self.distro = distro
+        self.instance_name = instance_name()
+
+    def find_user_data_file(self) -> Optional[PurePath]:
+        """
+        Finds the most precendent of the candidate files that may contain
+        user-data, if any, or None otherwise.
+        """
+        profile_dir = win_user_profile_dir()
+        seed_dir = os.path.join(profile_dir, ".cloud-init")
+        if not os.path.isdir(seed_dir):
+            LOG.error("%s directory doesn't exist.", seed_dir)
+            return None
+
+        # Notice that by default file name casing is irrelevant here. Windows
+        # filenames are case insensitive. Even though accessed through Linux,
+        # path translation just works with whichever casing we try.
+        # But users can change that behavior with configuration
+        # (ref https://learn.microsoft.com/en-us/windows/wsl/case-sensitivity),
+        # thus  better prevent it by always relying on case insensitive match.
+        existing_files = {
+            ef.name.casefold(): ef.path for ef in os.scandir(seed_dir)
+        }
+        if not existing_files:
+            LOG.error("%s directory is empty", seed_dir)
+            return None
+
+        folded_names = [
+            f.casefold()
+            for f in candidate_user_data_file_names(self.instance_name)
+        ]
+        for filename in folded_names:
+            if filename in existing_files.keys():
+                return PurePath(existing_files[filename])
+
+        LOG.error(
+            "%s doesn't contain any of the expected user-data files", seed_dir
+        )
+        return None
+
+    def _get_data(self) -> bool:
+        self.vendordata_raw = None
+
+        self.metadata = dict()
+        m_id = machine_id()
+        if m_id is None:
+            LOG.debug("Instance ID will be the WSL instance name only")
+            self.metadata["instance-id"] = self.instance_name
+        else:
+            self.metadata["instance-id"] = "{}-{}".format(
+                self.instance_name, m_id
+            )
+
+        try:
+            file = self.find_user_data_file()
```

Comment:
[Bump on my other comment in docs](https://github.com/canonical/cloud-init/pull/4786/files#r1462495371). Since it is a hard requirement that the WSL datasource has to minimally find at least one user-data file in order to use WSL images, I think the datasource documentation needs to express a default `config.user-data`  or `default.user-data` or `all.user-data` file needs to be present in order for cloud-init to be enabled on boot.